library/extractor/src/main/java/com/google/android/exoplayer2/text/dvb/DvbTeletextSubtitleParser.java

```diff
@@ -0,0 +1,511 @@
+  private final TeletextPage page = new TeletextPage();
+  private final ArrayList<Cue> cues = new ArrayList<>();
+
+  /**
+   * Construct an instance for the given magazine number and page number.
+   *
+   * @param magazineNumber The id of the subtitle magazine carrying the subtitle to be parsed.
+   * @param pageNumber The id of the subtitle page carrying the subtitle to be parsed.
+   */
+  public DvbTeletextSubtitleParser(int magazineNumber, int pageNumber) {
+    this.magazineNumber = magazineNumber;
+    this.pageNumber = pageNumber;
+  }
+
+  /** Resets the parser. */
+  public void reset() {
+    page.clear();
+    cues.clear();
+  }
+
+  /**
+   * Decodes a subtitling packet, returning a list of parsed {@link Cue}s.
+   *
+   * @param data The subtitling packet data to decode.
+   * @param size The limit in {@code data} at which to stop decoding.
+   * @return The parsed {@link Cue}s.
+   */
+  public List<Cue> decode(byte[] data, int size) {
+    scratch.reset(data, size);
+
+    int dataIdentifier = scratch.readUnsignedByte();
+
+    if (dataIdentifier < 0x10 || dataIdentifier > 0x1F) {
+      // ignore invalid data identifiers
+      return Collections.emptyList();
+    }
+
+    while (scratch.bytesLeft() > 2) {
+      int dataUnitId = scratch.readUnsignedByte();
+      int dataUnitLength = scratch.readUnsignedByte();
+
+      if (dataUnitLength > scratch.bytesLeft()) {
+        // ignore invalid data unit length
+        break;
+      }
+
+      switch (dataUnitId) {
+        case 0x02:
+          // EBU Teletext non-subtitle data
+          break;
+        case 0x03:
+          // EBU Teletext subtitle data
+          @Nullable TeletextLine line = processTeletextSubtitleField(scratch, dataUnitLength);
+          if (line != null) {
+            if (line.erase) {
+              cues.clear();
+            }
+            cues.add(buildCue(line));
+          }
+          break;
+        case 0xFF:
+          // stuffing
+          break;
+        default:
+          // ignore unknown data unit id
+          break;
+      }
+    }
+    return new ArrayList<>(cues);
+  }
+
+  private Cue buildCue(TeletextLine line) {
+    Cue.Builder cueBuilder = new Cue.Builder();
+    byte[] printable = new byte[TELETEXT_COLUMNS];
+    int color = TELETEXT_COLORS[7];
+    float size = (float) CHAR_HEIGHT / TELETEXT_HEIGHT;
+    boolean box = false;
+    boolean newBox = false;
+
+    for (int i = 0; i < line.line.length; ++i) {
+      int c = line.line[i] & 0xFF;
+      if (c < 0x20) {
+        if (c > 0 && c < 8) {
+          color = TELETEXT_COLORS[c];
+        } else if (c == 0xa) {
+          newBox = false;
+        } else if (c == 0xb) {
+          newBox = true;
+        } else if (c == 0xd) {
+          size = (float) CHAR_HEIGHT / TELETEXT_HEIGHT * 2;
+        }
+        c = ' ';
+      }
+
+      if (!box) {
+        c = ' ';
+      }
+      printable[i] = (byte) c;
+      box = newBox;
+    }
+
+    String latin1Line;
```

Comment:
nit: no need to name this `latin1Line`, can't it just be `line` or `cueText`?

You could even in-line it into the `SpannableString(Builder)` constructor on L369 imo to avoid having to name it at all.